hashintel/hash
# Realtime updates

## 🌟 What is the purpose of this PR?



This PR adds back `realtime` through postgres WAL (+ wal2json) and redis using the Graph database.
Write-ahead log logical replication is still the used way to subscribe to DB changes, as it offers most flexibility. `PG_NOTIFY` would require very manual setup and coupling to every table we subscribe to, whereas WAL replication with WAL2JSON is very flexible and wouldn't break with altered tables (only our downstream type mappings).

A couple changes had to be made to accommodate the realtime updates, see below list.

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1202787635216887/1202912971186621/f) _(internal)_

## 🚫 Blocked by




## 🔍 What does this change?



- Uses custom Dockerfile instead of `postgres:14-alpine3.16`. This postgres Dockerfile has to build WAL2JSON which allows us to interpret the write-ahead-log of postgres as JSON isntead of a binary format.
- Adds a news postgres user used exclusively for logical replication of the WAL log.
- Fixes existing realtime package to make use of the new Graph table structure.
- Slightly rewrites some of the connection logic in the realtime package.

## 📜 Does this require a change to the docs?



- No user-facing changes

## ⚠️ Known issues


- We've removed aggregations, so that subscription target is no longer usable/valid and has been commented out in collab. This means that collab won't compile, but lints will be unhappy.

## 🐾 Next steps


- Work on turning on collab again [Asana task](https://app.asana.com/0/1202805690238892/1202924026802709/f) _(internal)_

## 🛡 What tests cover this?



- None :(

## ❓ How to test this?



You might have to wipe your volumes because of the change of postgres images:
- `yarn external-services down`
- `docker volume prune`

Then run by:
1.  Checkout the branch
1.  Start external services `yarn external-services up --build`
1.  In `/packages/hash/realtime/` run `yarn dev`
1.  Observe a `Acquired slot ownership` info log
1.  If you have `redis-cli` starting it against the docker-instance of redis, 
     -  Run full `rest-test.http` file either with HTTPYac or Jetbrains IDE
     -  Open `redis-cli` and run `lrange collab 0 -1`

## 📹 Demo


![image](https://user-images.githubusercontent.com/6988668/188475863-c98bb14a-e013-4b01-acf5-8b5b732a3fdf.png)
